### ISSUE TITLE
RequireThis: xdoc examples don't produce violations like shown

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#4216]
Examples taken from http://checkstyle.sourceforge.net/config_coding.html#RequireThis .

https://github.com/checkstyle/checkstyle/blob/db64216a19b480855058d85f489d22fface4ff13/src/xdocs/config_coding.xml#L3472
````
$ cat TestClass.java
public static class B {
  private int field1;

  public A(int f) {
    field1 = f;
  }

  String addSuffixToField(String field1) {
    // Overlapping by method argument. Equal to "return field1 = field1 + "suffix";"
    return field1 += "suffix"; // violation: Reference to instance variable "field1" needs "this".
  }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="RequireThis"/>
    </module>
</module>

$ java -jar checkstyle-7.6.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

https://github.com/checkstyle/checkstyle/blob/db64216a19b480855058d85f489d22fface4ff13/src/xdocs/config_coding.xml#L3522
````
$ cat TestClass.java
public static class A {
  private int field1;
  private int field2;

  public A(int field1) {
    field1 = field1; // violation: Reference to instance variable "field1" needs "this".
    field2 = 0; // violation: Reference to instance variable "field2" needs "this".
    String field2;
    field2 = "0"; // No violation. Local var allowed
  }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="RequireThis">
    <property name="checkMethods" value="false"/>
</module>
    </module>
</module>

$ java -jar checkstyle-7.6.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:6:5: Reference to instance variable 'field1' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
````
Violation missing on line 7 for `field2 = 0;`.

https://github.com/checkstyle/checkstyle/blob/db64216a19b480855058d85f489d22fface4ff13/src/xdocs/config_coding.xml#L3539
````
$ cat TestClass.java
public static class B {
  private int field1;

  public A(int f) {
    field1 = f; // violation:  Reference to instance variable "field1" needs "this".
  }

  String addSuffixToField(String field1) {
    return field1 += "suffix"; // violation: Reference to instance variable "field1" needs "this".
  }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="RequireThis">
    <property name="checkMethods" value="false"/>
</module>
    </module>
</module>

$ java -jar checkstyle-7.6.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

All examples should be re-producible.
It should be made clear which config is being used by example.

**Comment by romani** [comment]
@AkshitaKukreja30 , please help us with this issue.

**Comment by AkshitaKukreja30** [comment]
I'm on it.

**Comment by romani** [comment]
@wate123 , please help us with this issue.

**Comment by wate123** [comment]
I need some hint. what's the TestConfig.xml and TestClass.java come from?

**Comment by wate123** [comment]
what are the TestConfig.xml TestClass.java come from? please tell how can I get the result?

**Comment by rnveach** [comment]
> what are the TestConfig.xml TestClass.java come from?

@wate123 They are nothing more than files that hold the information from the xdocs to show that the examples with the configurations given do not produce violations like they said they should. They could have been named anything, I just happen to use those names.
[We request all issues reported be reproducible by the CLI](http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug), which is why I gave those 2 files.

> please tell how can I get the result?

Just make new files from the examples I gave and run the CLI with the command I gave. The point of the full example is so we can reproduce the issue easily.

**Comment by achaJackson** [comment]
I'm on it.

**Comment by DXTkastb** [comment]
I'm on it.

**Comment by DXTkastb** [comment]
@romani  @strkkk 
Unfortunately I couldnt find an example where CLI with config:
```
<module name="RequireThis">
  <property name="checkMethods" value="false"/>
</module>
```
will throw violations.

The examples I have tried show violations for this config only when I set `validateOnlyOverlapping` to false.
I will make the changes in examples provided in the documentation to show correct violations. 
The only issue is with the above-mentioned config, please suggest some code that might work out.

**Comment by bossever** [comment]
@romani I think this issue can be closed in scope of pull request https://github.com/checkstyle/checkstyle/pull/9246.
The examples mentioned in the updated xdocs work as demonstrated, and ambiguous code has been removed.

**Comment by strkkk** [comment]
@bossever Thanks for pointing on this. Looks like docs where updated and issue is not relevant anymore.

**Comment by rnveach** [comment]
@strkkk I was able to reproduce the problem in the first example that I gave above. To me this issue still looks valid.

https://checkstyle.org/config_coding.html#RequireThis_Examples

````
$ cat TestClass.java
public static class A {
  private int field1;
  private int field2;

  public A(int field1) {
    // Overlapping by constructor argument.
    field1 = field1; // violation: Reference to instance variable "field1" needs "this".
    field2 = 0;
  }

  void foo3() {
    String field1 = "values";
    // Overlapping by local variable.
    field1 = field1; // violation:  Reference to instance variable "field1" needs "this".
  }
}

public static class B {
  private int field;

  public A(int f) {
    field = f;
  }

  String addSuffixToField(String field) {
    // Overlapping by method argument. Equal to "return field = field + "suffix";"
    return field += "suffix"; // violation: Reference to instance variable "field" needs "this".
  }
}

public static record C(int x) {
    void getTwoX(int x) {
        // Overlapping by method argument.
        return x += x; // violation: Reference to instance variable "x" needs "this".
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="RequireThis">
  <property name="checkMethods" value="false"/>
</module>
    </module>
</module>

$ java -jar checkstyle-8.40-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:7:5: Reference to instance variable 'field1' needs "this.". [RequireThis]
[ERROR] TestClass.java:14:5: Reference to instance variable 'field1' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 2 errors.
````

Line 27 and 34 are saying there should be violations, but none are produced. From what I read from XDoc, I reproduced the configuration as they specified.

Please reproduce the examples in CLI and share them to show if they do/don't produce the violations.

**Comment by 524130120** [comment]
@rnveach 
I reproduce the examples that don't produce the violations as expected.
And I find it is because the variable 'field' in 'return field += "suffix";' is regarded as a returned value. 
It happens in the method **getClassFrameWhereViolationIsFound**.

But, it is mentioned in the document that check will not raise violation for returned variable.
![requireThis](https://user-images.githubusercontent.com/67037143/164890001-c5b01c77-df08-4b14-bb8c-d0bd7346c37d.png)

So I wonder if this problem need to be fixed?
Thanks.

**Comment by romani** [comment]
@524130120 , please update documentation to remove `// violation: Reference to instance...`  as it should be ok. We cal leave a comment `ok as it is method argument override class field `

**Comment by romani** [comment]
Fixed by https://github.com/checkstyle/checkstyle/pull/15908
